### PR TITLE
Add Piccolo API components

### PR DESCRIPTION
**Pull request by dantownsend**
**Checklist**

- [x] This project is explicitly related to ASGI.
- [x] The new list entry contains a project name, URL and description.
- [x] Alphabetical order is preserved across entries and sections.

**What is this project?**

[Piccolo API](https://github.com/piccolo-orm/piccolo_api) - Provides session auth, token auth, rate limiting, and CSRF protection for ASGI apps.

**Do you know about other similar projects?**

No

**If so, how is this one different?**

---

_Anyone who agrees with this pull request can add a 👍._

### CODE REVIEW

**Comment by dantownsend** [comment]
Thanks Florimond.

Yeah, it's basically a collection of middleware and endpoints I needed to make an ASGI app production ready. I considered breaking it down into smaller packages, so for example the rate limiting middleware could be the `asgi_ratelimiting` package, but eventually there's PyPI fatigue, maintaining so many separate packages.

I think it's a good idea to list the relevant parts under separate headings. I'll amend the pull request tomorrow.

Good work with this project - it's a very useful resource.

**Comment by dantownsend** [comment]
Took another shot at it.

File: README.md [hunk 1]
```diff
@@ -125,7 +125,12 @@ _Tutorials about working with ASGI applications and components._
...
+- [CSRFMiddleware](https://piccolo-api.readthedocs.io/en/latest/csrf/index.html) - Protect against CSRF attacks when using cookies for authentication. (Shipped with Piccolo API.)
+- [CSPMiddleware](https://piccolo-api.readthedocs.io/en/latest/csp/index.html) - Tell browsers to only run Javascript from the same origin. (Shipped with Piccolo API.)
 - [HTTPSRedirectMiddleware](https://www.starlette.io/middleware/#httpsredirectmiddleware) - Redirect HTTP/WS traffic to HTTPS/WSS. (Shipped with Starlette.)
+- [RateLimitingMiddleware](https://piccolo-api.readthedocs.io/en/latest/rate_limiting/index.html) - Protect sensitive endpoints from brute force attacks. (Shipped with Piccolo API.)
+- [SessionAuth](https://piccolo-api.readthedocs.io/en/latest/session_auth/index.html) - Authenticate using sessions and cookies. (Shipped with Piccolo API, and requires Piccolo ORM.)
+- [TokenAuth](https://piccolo-api.readthedocs.io/en/latest/token_auth/index.html) - Authenticate using tokens in request headers. (Shipped with Piccolo API, and requires Piccolo ORM.)
```
Comment (florimondmanca) [inline README.md:133]: Would these be better suited to the [Authentication](https://github.com/florimondmanca/awesome-asgi#authentication) section?

File: README.md [hunk 2]
```diff
@@ -125,7 +125,12 @@ _Tutorials about working with ASGI applications and components._
 _Packages and components for managing the security of ASGI web applications._
 
 - [CORSMiddleware](https://www.starlette.io/middleware/#corsmiddleware) - Allow cross-origin requests from browsers. (Shipped with Starlette.)
+- [CSRFMiddleware](https://piccolo-api.readthedocs.io/en/latest/csrf/index.html) - Protect against CSRF attacks when using cookies for authentication. (Shipped with Piccolo API.)
+- [CSPMiddleware](https://piccolo-api.readthedocs.io/en/latest/csp/index.html) - Tell browsers to only run Javascript from the same origin. (Shipped with Piccolo API.)
```
Comment (florimondmanca) [inline README.md:129]: (Nit) P comes before R :-)

```suggestion
- [CSPMiddleware](https://piccolo-api.readthedocs.io/en/latest/csp/index.html) - Tell browsers to only run Javascript from the same origin. (Shipped with Piccolo API.)
- [CSRFMiddleware](https://piccolo-api.readthedocs.io/en/latest/csrf/index.html) - Protect against CSRF attacks when using cookies for authentication. (Shipped with Piccolo API.)
```

File: README.md (hunk 1 above)
Comment (dantownsend) [reply README.md]: Yeah, makes sense - updated.

**Pull request merged**